src/core/builder.rs

```diff
@@ -31,31 +34,94 @@ pub struct GraphBuilder {
     processing: HashSet<String>,
     verbose: bool,
     cache: Option<DepGraph>,
-    metadata: BuildMetadata
+    metadata: BuildMetadata,
+    // For profiling
+    pyro_server: Option<String>,
+    pyro_app_name: String,
+    pyro_invocation_count: u32,
 }
 
 #[pymethods]
 impl GraphBuilder {
     #[new]
-    pub fn new(verbose: Option<bool>) -> Self {
+    pub fn new(
+        verbose: Option<bool>,
+        pyro_server: Option<String>,
+        pyro_app_name: Option<String>,
+    ) -> Self {
         let verbose = if let Some(verbose) = verbose {
             verbose
         } else {
             false
         };
 
+        let pyro_app_name = if let Some(pyro_app_name) = pyro_app_name {
+            pyro_app_name
+        } else {
+            "fast-dep".to_string()
+        };
+
         let builder = GraphBuilder {
             graph: DepGraph::new(),
             processing: HashSet::new(),
             verbose: verbose,
             cache: None,
             metadata: BuildMetadata::new(),
+            // For profiling
+            pyro_server: pyro_server,
+            pyro_app_name: pyro_app_name,
+            pyro_invocation_count: 0,
         };
 
         builder
     }
 
     pub fn build(&mut self, source: &str, package: Option<String>) -> DepGraph {
+        let profiling_agent = if let Some(pyro_server) = self.pyro_server.clone() {
+            let session_name = format!("{}-{}", self.pyro_app_name, 
+            self.pyro_invocation_count);
+            info!(
+                "Building pyroscope agent: {}",
+                session_name
+            );
+            let agent = PyroscopeAgent::builder(
+                pyro_server,
+                session_name.clone()
+            ).backend(pprof_backend(PprofConfig::new().sample_rate(100)))
+            .build();
+
+            let agent = match agent {
+                Ok(agent) => Some(agent),
+                Err(error) => {
+                    warn!("Pyroscope agent build failed with error: {}", error);
+                    None
+                }
+            };
+
+            self.pyro_invocation_count += 1;
+
+            match agent {
+                Some(agent) => {
+                    match agent.start() {
+                        Ok(agent) => {
```

Comment:
I also feel this tends to happen, a lot of nested matching. I love matching and hate that Python / C++ does not have them but when I have like 3+ matches, I start to go dizzy. Probably this one is especially cause I re-use `agent` here a bunch, but the alternative is to have a bunch of extra vars floating around.

Reply:
You can reduce the match layers by using the `let Some() else {` pattern and moving the agent creation to its own function:

```
fn start_agent() {

  // code above to get initial agent value

  let Some(agent) = agent else {
    return None;
  }

  let Ok(agent) = agent.start() else {
    // error handling
    return None;
  }

  // return started agent
  Some(agent)
}
```